### PR TITLE
Use controlled components in tutorial

### PR DESCRIPTION
**Pull request by yangshun**
Updated tutorial and use controlled components instead of refs, as mentioned in #5443. @zpao

### CODE REVIEW

File: docs/docs/tutorial.md [hunk 1]
```diff
@@ -494,26 +494,35 @@ var CommentForm = React.createClass({
...
-    return;
+    this.setState({author: '', text: ''});
   },
   render: function() {
     return (
       <form className="commentForm" onSubmit={this.handleSubmit}>
-        <input type="text" placeholder="Your name" ref="author" />
-        <input type="text" placeholder="Say something..." ref="text" />
+        <input type="text" placeholder="Your name"
+          value={this.state.author} onChange={this.handleAuthorChange}/>
```
Comment (zpao) [inline docs/docs/tutorial.md]: I know this is a lot of lines but its our style and I think we should use it so we're consistent with the rest of our code samples.

```js
<input
  type="text"
  placeholder="Your name"
  value={this.state.author}
  onChange={this.handleChange}
/>
```

**Comment by zpao** [comment]
Thanks for taking this on!

This is a lot of code to introduce all at once - I think we might need a bit more in the paragraph preceding that first big code change to tie it together more.

> Let's make the form interactive. When the user submits the form, we should clear it, submit a request to the server, and refresh the list of comments. To start, let's listen for the form's submit event and clear it.

It's technically accurate - that is exactly what we're doing. But then you see a whole lot of code changing to do something that on the surface seems pretty straightforward. So I think saying something about the fact that we're going to use `state` to store the user input as it is entered - another use of state that we didn't discuss above when we talked about state. We're also definitely doing more than "listen for the form's submit event and clear it".

**Comment by yangshun** [comment]
@zpao Great suggstions! I have broken the code changes down into smaller and more digestable chunks. Have a read of the modified version [here](http://yangshun.im/react/docs/tutorial.html#adding-new-comments)

**Comment by yangshun** [comment]
@zpao what do you think?

File: docs/docs/tutorial.md [hunk 2]
```diff
@@ -492,53 +492,108 @@ var CommentForm = React.createClass({
...
 
-Let's make the form interactive. When the user submits the form, we should clear it, submit a request to the server, and refresh the list of comments. To start, let's listen for the form's submit event and clear it.
+#### Controlled components
+
+With the traditional DOM, `input` elements are rendered and the browser manages the state (its rendered value). As a result, the state of the actual DOM will differ from that of the component. This is not ideal as the state of the view will differ from that of the component. In React, components should always represent the state of the view and not only at the point of initialization.
+
+Hence, we will be using `this.state` to save the user's input as it is entered. We define an initial `state` with two properties `author` and `text` and set them to be empty strings. In our `<input>` elements, we set the `value` prop to reflect the `state` of the component and attach `onChange` handlers to them. These `<input>` elements with a `value` set are called controlled components. Read more about controlled components on the [Forms article](/react/docs/forms.html#controlled-components).
 
-```javascript{3-14,16-19}
+```javascript{3-11,14-26}
```
Comment (zpao) [inline docs/docs/tutorial.md]: Could you make this 27 (to catch the closing `</form>` tag (it's already messed up on the site so not your fault)

**Comment by zpao** [comment]
Otherwise, this looks really awesome, thanks! Could you fix that one bit and squash the commits, then we can merge this. Would you like to update the tutorial repo as well (https://github.com/reactjs/react-tutorial/blob/master/public/scripts/example.js)? I'm happy to do that but figured I'd give you the chance if you wanted :)

**Comment by yangshun** [comment]
Yes I'd gladly make the fixes for the tutorial repo too! Gimme a moment to modify the highlighting and submit a PR to the tutorial repo (:

**Comment by yangshun** [comment]
@zpao Do I have to update the tutorials of the other languages too?

**Comment by zpao** [comment]
No, let's let the translators do that. Unless you happen to know Italian, Japanese, Korean, and Chinese…

**Comment by yangshun** [comment]
But they will have to sync up the code changes too since another snippet has been added. Will that be fine?

**Comment by zpao** [comment]
Yea, that'll be fine. We don't do a good job of surfacing those translations so it's not a huge deal.

**Comment by zpao** [comment]
Awesome, let's get these in and I'll update the site :)

**Comment by yangshun** [comment]
Got it! The PR for the react-tutorial can be found here: https://github.com/reactjs/react-tutorial/pull/101

**Pull request merged**